kratos/includes/registry_item.h

```diff
@@ -181,23 +199,69 @@ class KRATOS_API(KRATOS_CORE) RegistryItem
     ///@{
 
     // Adding sub value item
-    template<typename TItemType, class... TArgumentsList>
-    RegistryItem& AddItem(
-        std::string const& ItemName,
-        TArgumentsList&&... Arguments)
+    template <typename TItemType, class... TArgumentsList>
+    RegistryItem &AddItem(
+        std::string const &ItemName,
+        TArgumentsList &&...Arguments)
     {
         KRATOS_ERROR_IF(this->HasItem(ItemName))
             << "The RegistryItem '" << this->Name() << "' already has an item with name "
             << ItemName << "." << std::endl;
 
-        using ValueType = typename std::conditional<std::is_same<TItemType, RegistryItem>::value, SubRegistryItemFunctor, SubValueItemFunctor<TItemType>>::type;
+        if constexpr (std::is_same_v<TItemType, RegistryItem>) {
+            using ValueType = typename std::conditional<std::is_same<TItemType, RegistryItem>::value, SubRegistryItemFunctor, SubValueItemFunctor<TItemType>>::type;
+
+            auto insert_result = GetSubRegistryItemMap().emplace(
+                std::make_pair(
+                    ItemName,
+                    ValueType::Create(ItemName, std::forward<TArgumentsList>(Arguments)...)));
+
+            KRATOS_ERROR_IF_NOT(insert_result.second)
+                << "Error in inserting '" << ItemName
+                << "' in registry item with name '" << this->Name() << "'." << std::endl;
+
+            return *insert_result.first->second;
+        }
+        else
+        {
+            std::function<std::shared_ptr<TItemType>()> aux_lambda = [=]() -> std::shared_ptr<TItemType> {
+                return std::make_shared<TItemType>((Arguments)...);
```

Comment:
```suggestion
                return std::make_shared<TItemType>(Arguments...);
```

ah yes, C(lojure)++, my favorite language :D

Reply:
Solved, I will push with the rest of the changes :+1: